groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+
+template <class TYPE = VARIANT_NOT_A_TYPE, class... OTHER>
+struct UsesBslmaAllocatorAny;
+
+template <>
+struct UsesBslmaAllocatorAny<VARIANT_NOT_A_TYPE> : bsl::false_type {
+};
+
+template <class TYPE, class... OTHER>
+struct UsesBslmaAllocatorAny
+: bsl::integral_constant<
+      bool,
+      BloombergLP::bslma::UsesBslmaAllocator<TYPE>::value ||
+          BloombergLP::bslma::UsesBslmaAllocatorAny<OTHER...>::value> {
+    // TODO : move to bslma_usesbslmaallocator or rename as
+    // Variant_UsesBslmaAllocatorAny ?
+};
+
+template <class TYPE = VARIANT_NOT_A_TYPE, class... OTHER>
+struct IsBitwiseMoveableAll;
+
+template <>
+struct IsBitwiseMoveableAll<VARIANT_NOT_A_TYPE> : bsl::false_type {
+};
+
+template <class TYPE, class... OTHER>
+struct IsBitwiseMoveableAll
+: bsl::integral_constant<bool,
+                         BloombergLP::bslmf::IsBitwiseMoveable<TYPE>::value &&
+                             IsBitwiseMoveableAll<OTHER...>::value> {
+    // TODO : move to another component or rename as
+    // Variant_IsBitwiseMoveableAll ?
+};
+
+#endif
+
+}  // close package namespace
+}  // close enterprise namespace
+
+// helper classes
+namespace bsl {
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+template <class HEAD, class... TAIL>
+struct variant;
+#endif
+
+// Return value from `index()` method if 'valueless_by_exception()' is 'true'
+BSLS_KEYWORD_INLINE_CONSTEXPR size_t variant_npos = -1;
+
+template <size_t INDEX, class TYPE>
+struct variant_alternative;  // primary template not defined
+
+template <size_t INDEX, class TYPE>
+struct variant_alternative<INDEX, const TYPE> {
+    // This metafunction applies the cv qualifier of the 'variant' type to its
+    // alternative types.
+    typedef typename variant_alternative<INDEX, TYPE>::type const type;
+};
+
+template <size_t INDEX, class TYPE>
+struct variant_alternative<INDEX, volatile TYPE> {
+    // This metafunction applies the cv qualifier of the 'variant' type to its
+    // alternative types.
+    typedef typename variant_alternative<INDEX, TYPE>::type volatile type;
+};
+
+template <size_t INDEX, class TYPE>
+struct variant_alternative<INDEX, const volatile TYPE> {
+    // This metafunction applies the cv qualifier of the 'variant' type to its
+    // alternative types.
+    typedef
+        typename variant_alternative<INDEX, TYPE>::type const volatile type;
+};
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+// 'Variant_VariantAlternativeImpl' defined to avoid 'variant<>' from the
+// autogeneration script
+template <size_t INDEX, class HEAD = VARIANT_NOT_A_TYPE, class... TAIL>
+struct Variant_VariantAlternativeImpl
+: Variant_VariantAlternativeImpl<INDEX - 1, TAIL...> {
+};
+
+template <size_t INDEX>
+struct Variant_VariantAlternativeImpl<INDEX, VARIANT_NOT_A_TYPE> {
+};
+
+template <class HEAD, class... TAIL>
+struct Variant_VariantAlternativeImpl<0, HEAD, TAIL...> {
+    typedef HEAD type;
+};
```

Comment:
I think this also works and is simpler:
```suggestion
template <size_t INDEX, class... TYPES>
struct Variant_VariantAlternativeImpl {
};

template <size_t INDEX, class HEAD, class... TAIL>
struct Variant_VariantAlternativeImpl<INDEX, HEAD, TAIL...>
: Variant_VariantAlternativeImpl<INDEX - 1, TAIL...> {
};

template <class HEAD, class... TAIL>
struct Variant_VariantAlternativeImpl<0, HEAD, TAIL...> {
    typedef HEAD type;
};
```

Reply:
closing this as it is needed (discussion in a different conversation), comment done in `VARIANT_NOT_A_TYPE`